### PR TITLE
Saving compiled code

### LINKED ISSUE #13: Saving compiled code from cfunction objects in user defined locations is not supported
I frequently generate cfunctions in my mkin package by generating C code from a user specified ODE model definition which is then compiled for fast solutions with deSolve within iterative optimizations. Generally, it is fine to have the compiled code under /tmp, but when I have large knitr documents and specific chunks take a long time to complete I use knitr chunk caching and I get problems with model objects in the knitr cache not finding their compiled code any more. Therefore, I would like to store the shared objects generated by cfunctions in a user defined path, e.g. in the current working directory.

The functions readDynLib and writeDynLib from R/utilities.R would fit the bill, but they are not exported, and the former has a bug making the example code in dontrun fail, even if the functions are found by using `inline:::readDynLib` and `inline:::writeDynLib`. An alternative to fixing and exporting these functions would be to add an argument to cfunction, defining a path for the shared object.

### PR DESCRIPTION
**Pull request by jranke**
This PR adds functionality for writing CFunc objects. The previously existing code defining `writeDynLib` and `readDynLib` was not functional for several reasons. Therefore, a new approach was developed (see commit messages for details), working in two steps.

A precondition for this to work is that the finalizer created by `cfunction` does not remove the DLL that `libLFile` in the environment of the function points to. Therefore, `cfunction` was modified to only remove `libLFile_orig` upon garbage collection or session termination, but to leave `libLFile` which will point to a different location after using `moveDLL`.

`moveDLL` moves the DLL to a user defined location for an existing CFunc object, i.e. it modifies its argument. Having a separate step for this allows for explicit control over unloading and overwriting of any pre-existing DLLs. Care is taken that no loaded DLL gets overwritten, which would cause a segfault.

`writeCFunc` writes a `CFunc` object, but takes care that the DLL has been moved already, as otherwise it would be garbage collected, making the operation pointless. Side note: We could also recompile when reading...

`readCFunc` reads a `CFunc` object stored using `writeCFunc` and restores the pointer (symbol info) in the function body.

`cfunction` also gains an argument `name` that will be used as the function name in the source code if the signature is not specified as a named list (sorry for not splitting out this as a separate issue with a separate PR).

And please excuse the whitespace changes in the docs (`man/utilities.R`). I have a hard time checking .Rd files when indentation and whitespace is too random.

All the new functionality is tested in example code and tiny tests.

Closes #13.

### CODE REVIEW

**Comment by eddelbuettel** [comment]
~I cannot merge this as is as it changes published interfaces.~

My bad. Apparently only internal functions.

Would you consider rebasing or would you mind squashing?  This is one PR.  I am not sure we need 30 or so commits here.

**Comment by jranke** [comment]
I don't mind squashing at all.

File: DESCRIPTION [hunk 1]
```diff
@@ -1,9 +1,9 @@
 Package: inline
 Version: 0.3.16.1
-Date: 2020-11-25
+Date: 2020-11-26
```
Comment (eddelbuettel) [inline DESCRIPTION:3]: :grinning: 

I usually leave version micro-variants and date for releases. (Also, I just made it 0.3.16.2 here on the checkout of the PR to make it distinct, but I prefer if these are not touched by PRs, so all good.)

File: DESCRIPTION [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
 Title: Functions to Inline C, C++, Fortran Function Calls from R
 Author: Oleg Sklyar, Duncan Murdoch, Mike Smith, Dirk Eddelbuettel, 
- Romain Francois, Karline Soetaert
+ Romain Francois, Karline Soetaert, Johannes Ranke
```
Comment (eddelbuettel) [inline DESCRIPTION:6]: Nice.

File: NAMESPACE [hunk 3]
```diff
@@ -5,14 +5,17 @@ export(
     "cfunction",
     "cxxfunction",
     "getPlugin",
+    "readCFunc",
     "registerPlugin",
-    "rcpp"
+    "rcpp",
+    "writeCFunc"
```
Comment (eddelbuettel) [inline NAMESPACE:11]: Do these actually persist C-level functions across sessions now?  We were never clear if that could be done.

**Review by eddelbuettel — APPROVED** [review]
Nice and careful work. Good to have the tests.

**Comment by eddelbuettel** [comment]
I added one quick commit to master to update NEWS (and roll the minor). If you have more work planned please pull and rebase as needed.

**Comment by eddelbuettel** [comment]
(And two more)

File: DESCRIPTION (hunk 1 above)
Comment (jranke) [reply DESCRIPTION:3]: :+1:

File: NAMESPACE (hunk 3 above)
Comment (jranke) [reply NAMESPACE:11]: Yes, that's exactly the point. I just ran

```
x <- as.numeric(1:10)
n <- as.integer(10)

code <- "
      integer i
      do 1 i=1, n(1)
    1 x(i) = x(i)**3
"
cubefn <- cfunction(signature(n="integer", x="numeric"), code,
  convention=".Fortran")
code(cubefn)

cubefn(n, x)$x

moveDLL(cubefn, name = "cubefn", directory = "~/dll")
path <- file.path("~/dll", "cubefn.rda")
writeCFunc(cubefn, path)
```

in one session and

```
path <- file.path("~/dll", "cubefn.rda")
cfn <- readCFunc(path)
cfn(3, 1:3)$x
```

and got

```
> cfn(3, 1:3)$x
[1]  1  8 27
```

using 0.3.16.2 on Linux

`\o/`

**Pull request merged**